File: 2025/day1.py
from utils import run_day


def parse_input(txt: str) -> list[int]:
    """Parse input, storing left rotations as negative numbers and right
    rotations as positive numbers"""
    rotations = []
    for line in txt.splitlines():
        if not line:
            continue
        num = int(line[1:])
        if line[0] == "L":
            num = -num
        rotations.append(num)

    return rotations


def solve_part1(rotations: list[int]) -> int:
    num_zeroes = 0
    position = 50
    for rotation in rotations:
        position = (position + rotation) % 100
        if position == 0:
            num_zeroes += 1

    return num_zeroes


def solve_part2(rotations: list[int]) -> int:
    num_zeroes = 0
    position = 50
    for rotation in rotations:
        # count how many times we pass zero rather than just how many times we land on it
        if rotation > 0:
            num_zeroes += (position + rotation) // 100
        else:
            num_zeroes += ((100 - position) % 100 - rotation) // 100
        position = (position + rotation) % 100

    return num_zeroes


if __name__ == "__main__":
    run_day("day1-input.txt", parse_input, solve_part1, solve_part2)

File: 2025-exs/day1.exs
defmodule Day1 do
  @doc """
  Parse input to a list, with left rotations as negative numbers and right rotations
  as positive numbers.
  """
  def parse_input(txt) do
    for line <- String.split(txt),
        <<dir, num_str::binary>> = line,
        num = String.to_integer(num_str) do
      if dir == ?L do
        -num
      else
        num
      end
    end
  end

  def solve_part1(rotations) do
    {num_zeros, _} =
      Enum.reduce(rotations, {0, 50}, fn rotation, {num_zeros, position} ->
        position = Integer.mod(position + rotation, 100)

        {if position == 0 do
           num_zeros + 1
         else
           num_zeros
         end, position}
      end)

    num_zeros
  end

  def solve_part2(rotations) do
    {num_zeros, _} =
      Enum.reduce(rotations, {0, 50}, fn rotation, {num_zeros, position} ->
        num_zeros =
          if rotation > 0 do
            num_zeros + div(position + rotation, 100)
          else
            num_zeros + div(Integer.mod(100 - position, 100) - rotation, 100)
          end

        position = Integer.mod(position + rotation, 100)
        {num_zeros, position}
      end)

    num_zeros
  end
end

Aoc2025.Utils.run_day("day1-input.txt", &Day1.parse_input/1, &Day1.solve_part1/1, &Day1.solve_part2/1)
